src/ui/ui_grid.cpp

```diff
@@ -0,0 +1,123 @@
+#include "ui_grid.h"
+
+#include "util/colors.h"
+#include "events.h"
+
+UIGrid::~UIGrid() {
+    for (unsigned int i = 0; i < children.size(); ++i)
+        delete children[i];
+}
+
+UIGrid::UIGrid(float start_x, float start_y,
+        int num_elements, int columns,
+        float element_width, float element_height,
+        glm::vec3 grid_bg_color,
+        float border_width,
+        float inter_padding,
+        float selection_halo_padding)
+    : UIContainer(start_x, start_y),
+    num_elements(num_elements), columns(columns),
+    element_width(element_width), element_height(element_height),
+    grid_bg_color(grid_bg_color),
+    border_width(border_width), inter_padding(inter_padding),
+    selection_halo_padding(selection_halo_padding),
+    selection_row(0), selection_col(0) {
+
+    // Calculate number of rows.
+    rows = (num_elements - 1) / columns + 1;
+    // Calculate grid dimension.
+    grid_height = 2 * border_width
+                  + rows * element_height
+                  + (rows - 1) * inter_padding;
+    grid_width = 2 * border_width
+                 + columns * element_width
+                 + (columns - 1) * inter_padding;
+    // Size of elements including padding
+    total_element_width = element_width + inter_padding;
+    total_element_height = element_height + inter_padding;
+
+    // Create grid background element and add to children
+    grid_bg = UIRectangle(start_x, start_y,
+                          grid_width, grid_height,
+                          grid_bg_color);
+
+    // Build empty attachment points across the grid.
+    float elt_start_x = start_x + border_width;
+    float elt_start_y = start_y + grid_height - border_width - element_height;
+    int curr_elt = 0;
+    for (unsigned int row = 0; row < rows; ++row) {
+        for (unsigned int col = 0; col < columns; ++col) {
+            float adjusted_x = elt_start_x + (total_element_width * col);
+            float adjusted_y = elt_start_y - (total_element_height * row);
+
+            UIHaloContainer *attach_point =
+                new UIHaloContainer(adjusted_x, adjusted_y,
+                    element_width, element_height, selection_halo_padding);
+            children.push_back(attach_point);
+
+            curr_elt++;
+            if (curr_elt == num_elements) break; // finished doing all the elements!
+        }
+    }
+
+    // First item in grid is selected by default (0,0)
+    toggle_current_selection_halo();
+
+    // Set up callbacks.
+    // TODO: add semantic game logic abstraction layer
+    events::joystick_event.connect([&](events::JoystickData d) {
+            move_selection(Direction::DOWN);});
```

Comment:
I'm assuming this is just for test/debugging?  Seems like you'd want to move in a direction based on the joystick input, rather than just always down.